diff --git a/dist/index.d.mts b/dist/index.d.mts
index 98615cb6b357e2489173045dd6aa266d76dff1e4..41a97de8c05991bf47cc3b523f252f57ff6d3d78 100644
--- a/dist/index.d.mts
+++ b/dist/index.d.mts
@@ -10,5 +10,6 @@ interface Options {
 }
 
 declare const convert: <T extends object = JSONSchema4>(schema: T, options?: Options) => Promise<OpenAPIV3.Document>;
+declare const convertSync: <T extends object = JSONSchema4>(schema: T, options?: Options) => OpenAPIV3.Document;
 
-export { convert, convert as default };
+export { convert, convertSync, convert as default };
diff --git a/dist/index.d.ts b/dist/index.d.ts
index 98615cb6b357e2489173045dd6aa266d76dff1e4..41a97de8c05991bf47cc3b523f252f57ff6d3d78 100644
--- a/dist/index.d.ts
+++ b/dist/index.d.ts
@@ -10,5 +10,6 @@ interface Options {
 }
 
 declare const convert: <T extends object = JSONSchema4>(schema: T, options?: Options) => Promise<OpenAPIV3.Document>;
+declare const convertSync: <T extends object = JSONSchema4>(schema: T, options?: Options) => OpenAPIV3.Document;
 
-export { convert, convert as default };
+export { convert, convertSync, convert as default };
diff --git a/dist/index.js b/dist/index.js
index 6d269a3d4437adbfd8466b28cbc5387800554d39..0102eacf6eba4c9a32a25bb6099340b6b086acc3 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -21,6 +21,7 @@ var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: tru
 var index_exports = {};
 __export(index_exports, {
   convert: () => convert,
+  convertSync: () => convertSync,
   default: () => index_default
 });
 module.exports = __toCommonJS(index_exports);
@@ -120,6 +121,43 @@ var handleDefinition = async (def, schema) => {
   }
   return def;
 };
+var handleDefinitionSync = (def, schema) => {
+  if (typeof def !== "object") {
+    return def;
+  }
+  const type = def.type;
+  if (type) {
+    const walker = new import_json_schema_walker.Walker();
+    walker.loadSchemaSync(
+      {
+        definitions: schema["definitions"] || [],
+        ...def,
+        $schema: schema["$schema"]
+      },
+      {
+        cloneSchema: true
+      }
+    );
+    walker.walkSync(convertSchema, walker.vocabularies.DRAFT_07);
+    if ("definitions" in walker.rootSchema) {
+      delete walker.rootSchema.definitions;
+    }
+    return walker.rootSchema;
+  }
+  if (Array.isArray(def)) {
+    const typeArr = def;
+    const hasNull = typeArr.includes("null");
+    if (hasNull) {
+      const actualTypes = typeArr.filter((l) => l !== "null");
+      return {
+        type: actualTypes.length === 1 ? actualTypes[0] : actualTypes,
+        nullable: true
+        // this is incorrect but thats ok, we are in the inbetween phase here
+      };
+    }
+  }
+  return def;
+};
 var convert = async (schema, options) => {
   const walker = new import_json_schema_walker.Walker();
   const convertDefs = options?.convertUnreferencedDefinitions ?? true;
@@ -134,6 +172,20 @@ var convert = async (schema, options) => {
   }
   return rootSchema;
 };
+var convertSync = (schema, options) => {
+  const walker = new import_json_schema_walker.Walker();
+  const convertDefs = options?.convertUnreferencedDefinitions ?? true;
+  walker.loadSchemaSync(schema, options);
+  walker.walkSync(convertSchema, walker.vocabularies.DRAFT_07);
+  const rootSchema = walker.rootSchema;
+  if (convertDefs && rootSchema?.definitions) {
+    for (const defName in rootSchema.definitions) {
+      const def = rootSchema.definitions[defName];
+      rootSchema.definitions[defName] = handleDefinitionSync(def, schema);
+    }
+  }
+  return rootSchema;
+};
 function stripIllegalKeywords(schema) {
   if (typeof schema !== "object") {
     return schema;
@@ -322,5 +374,6 @@ function rewriteExclusiveMinMax(schema) {
 var index_default = convert;
 // Annotate the CommonJS export names for ESM import in node:
 0 && (module.exports = {
-  convert
+  convert,
+  convertSync
 });
diff --git a/dist/index.mjs b/dist/index.mjs
index c75f758bfa31607702436f625eb6f7c22e9d6071..36cd5caf01a79b61698707ecf1b8bd63c4900ce3 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -95,6 +95,43 @@ var handleDefinition = async (def, schema) => {
   }
   return def;
 };
+var handleDefinitionSync = (def, schema) => {
+  if (typeof def !== "object") {
+    return def;
+  }
+  const type = def.type;
+  if (type) {
+    const walker = new Walker();
+    walker.loadSchemaSync(
+      {
+        definitions: schema["definitions"] || [],
+        ...def,
+        $schema: schema["$schema"]
+      },
+      {
+        cloneSchema: true
+      }
+    );
+    walker.walkSync(convertSchema, walker.vocabularies.DRAFT_07);
+    if ("definitions" in walker.rootSchema) {
+      delete walker.rootSchema.definitions;
+    }
+    return walker.rootSchema;
+  }
+  if (Array.isArray(def)) {
+    const typeArr = def;
+    const hasNull = typeArr.includes("null");
+    if (hasNull) {
+      const actualTypes = typeArr.filter((l) => l !== "null");
+      return {
+        type: actualTypes.length === 1 ? actualTypes[0] : actualTypes,
+        nullable: true
+        // this is incorrect but thats ok, we are in the inbetween phase here
+      };
+    }
+  }
+  return def;
+};
 var convert = async (schema, options) => {
   const walker = new Walker();
   const convertDefs = options?.convertUnreferencedDefinitions ?? true;
@@ -109,6 +146,20 @@ var convert = async (schema, options) => {
   }
   return rootSchema;
 };
+var convertSync = (schema, options) => {
+  const walker = new Walker();
+  const convertDefs = options?.convertUnreferencedDefinitions ?? true;
+  walker.loadSchemaSync(schema, options);
+  walker.walkSync(convertSchema, walker.vocabularies.DRAFT_07);
+  const rootSchema = walker.rootSchema;
+  if (convertDefs && rootSchema?.definitions) {
+    for (const defName in rootSchema.definitions) {
+      const def = rootSchema.definitions[defName];
+      rootSchema.definitions[defName] = handleDefinitionSync(def, schema);
+    }
+  }
+  return rootSchema;
+};
 function stripIllegalKeywords(schema) {
   if (typeof schema !== "object") {
     return schema;
@@ -297,5 +348,6 @@ function rewriteExclusiveMinMax(schema) {
 var index_default = convert;
 export {
   convert,
+  convertSync,
   index_default as default
 };
